File: Python Solutions/fenwick.py
def sum(tree, i):
    res = 0

    while i > 0:
        res += tree[i]
        i -= i & (-i)

    return res

def update(tree , N , i, val):
    i += 1

    while i <= N:
        tree[i] += val
        i += i & (-i)


def main():
    N,Q = map(int, input().split())
    ftTree = [0] * (N + 1)

    for i in range(Q):
        command = input().split()
        if(command[0] == "+"):
            update(ftTree, N, int(command[1]), int(command[2]))
        if(command[0] == "?"):
            print(sum(ftTree, int(command[1])))

main()

File: C++ Solutions/fenwick.cpp
#include <cstdio>

long long tree[5000005];

void update(int i, long long value, int N)
{
    while(i <= N)
    {
        tree[i] += value;
        i += i & -i;
    }
}
long long sum(int i)
{
     long long sum = 0;

     while (i > 0)
     {
        sum += tree[i];
        i -= i & (-i);
     }
     return sum;
}


int main()
{

    long long value;
    int N, Q, i;
    char c;
    scanf("%d%d", &N, &Q);

    for(int j = 0; j < Q; j++)
    {
        scanf(" %c ", &c);


        if (c == '+')
        {
            scanf("%d%lld", &i, &value);
            update(i + 1, value, N);
        }

        if (c == '?')
        {
            scanf("%d", &i);
            printf("%lld\n", sum(i));
        }
    }

    return 0;
}
